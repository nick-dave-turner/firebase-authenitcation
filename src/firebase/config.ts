/* istanbul ignore file */
interface FirebaseConfig {
  apiKey: string | undefined;
  appId: string | undefined;
  authDomain: string | undefined;
  databaseURL: string | undefined;
  measurementId: string | undefined;
  messagingSenderId: string | undefined;
  projectId: string | undefined;
  storageBucket: string | undefined;
}

export const firebaseConfig: FirebaseConfig = {
  apiKey: process.env.REACT_APP_API_KEY,
  appId: process.env.REACT_APP_APP_ID,
  authDomain: process.env.REACT_APP_AUTH_DOMAIN,
  databaseURL: process.env.REACT_APP_DATABASE_URL,
  measurementId: process.env.REACT_APP_MEASUREMENT_ID,
  messagingSenderId: process.env.REACT_APP_MESSAGING_SENDER_ID,
  projectId: process.env.REACT_APP_PROJECT_ID,
  storageBucket: process.env.REACT_APP_PROJECT_ID,
};
