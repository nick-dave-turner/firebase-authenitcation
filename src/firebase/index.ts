/* istanbul ignore file */
export { AuthProvider, useAuthContext } from "./context";
export { firebaseAuth } from "./init";
export { useAuth } from "./use-auth";
