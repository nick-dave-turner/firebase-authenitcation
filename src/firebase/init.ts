/* istanbul ignore file */
import firebase from "firebase/app";
import "firebase/auth";

import { firebaseConfig } from "./config";

export const firebaseApp = firebase.initializeApp(firebaseConfig);
export const firebaseAuth = firebaseApp.auth();
