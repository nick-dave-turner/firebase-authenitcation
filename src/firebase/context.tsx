import React from "react";

import { firebaseAuth } from "./init";

type AuthType = { user: firebase.User | undefined };

const AuthContext = React.createContext<AuthType>({ user: undefined });

export const AuthProvider: React.FC = ({ children }) => {
  const [user, setUser] = React.useState<AuthType>({ user: undefined });
  const [initializing, setInitializing] = React.useState<boolean>(true);

  React.useEffect(() => {
    const subscriber = firebaseAuth.onAuthStateChanged((result) => {
      setUser({ user: result || undefined });
      if (initializing) setInitializing(false);
    });

    return subscriber;
  }, [initializing]);

  if (initializing) return <h1>Loading...</h1>;

  return <AuthContext.Provider value={user}>{children}</AuthContext.Provider>;
};

export const useAuthContext = (): AuthType => React.useContext(AuthContext);
