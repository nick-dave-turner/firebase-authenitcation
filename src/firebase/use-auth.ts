import { useHistory } from "react-router-dom";
import { firebaseAuth } from "./init";

interface Error {
  error?: string;
}

interface AuthWithEmail {
  email: string;
  password: string;
}

interface UseAuth {
  signInWithEmail: ({ email, password }: AuthWithEmail) => Promise<Error>;
  signUpWithEmail: ({ email, password }: AuthWithEmail) => Promise<Error>;
  signOut: () => Promise<void>;
}

export const FAILED_MESSAGE =
  "Sorry, there was an error verifying your account details. Please try again.";

export const useAuth = (): UseAuth => {
  const history = useHistory();

  const signInWithEmail = ({
    email,
    password,
  }: AuthWithEmail): Promise<Error> => {
    return new Promise((resolve, reject) => {
      firebaseAuth
        .signInWithEmailAndPassword(email, password)
        .catch(() => reject(FAILED_MESSAGE));
    });
  };

  const signUpWithEmail = ({
    email,
    password,
  }: AuthWithEmail): Promise<{ error?: string }> => {
    return new Promise((resolve, reject) => {
      firebaseAuth
        .createUserWithEmailAndPassword(email, password)
        .catch(({ message }) => reject(message));
    });
  };

  const signOut = (): Promise<void> => {
    return firebaseAuth
      .signOut()
      .then(() => history.push("/signin"))
      .catch((error) => error);
  };

  return {
    signInWithEmail,
    signUpWithEmail,
    signOut,
  };
};
