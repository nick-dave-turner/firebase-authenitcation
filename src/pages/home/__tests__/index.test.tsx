import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import pretty from "pretty";

import { useAuth } from "../../../firebase";
import { HomePage } from "..";

jest.mock("../../../firebase");

const MOCK_SIGNOUT = jest.fn();

(useAuth as jest.Mock).mockImplementation(() => ({
  signOut: MOCK_SIGNOUT,
}));

describe("HomePage", () => {
  it("should match snapshot", () => {
    const { container } = render(<HomePage />);
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("should call signout when signout button clicked", () => {
    render(<HomePage />);
    fireEvent.click(screen.getByTestId("sign-out-button"));
    expect(MOCK_SIGNOUT).toHaveBeenCalled();
  });
});
