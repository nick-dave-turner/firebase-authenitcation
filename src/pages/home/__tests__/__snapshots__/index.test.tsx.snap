// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should match snapshot 1`] = `"<h1>Home Page</h1><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"button\\" data-testid=\\"sign-out-button\\"><span class=\\"MuiButton-label\\">Sign Out</span><span class=\\"MuiTouchRipple-root\\"></span></button>"`;
